### PR TITLE
Expand Support To Unbound

### PR DESCRIPTION
**Pull request by evanwporter**
This will expand support to Pokemon Unbound. This will be made possible with #80

### CODE REVIEW

**Comment by evanwporter** [comment]
We should make an abstract class for use by RR and UB. This will promote code reuse.

**Comment by evanwporter** [comment]
@andrewbenington 

Unbound is looking really good. Several issues however:

* Game of Origin
* Sprites. Unbound has more forms than RR (like all the forms of vivillion). Will need to selectively copy over the forms from [here](https://github.com/Skeli789/Dynamic-Pokemon-Expansion/tree/master/graphics/frontspr).
  * A way do this is to compare the current sprite list and the target sprite list (link above) and remove all duplicate files (names are different though so will need to read file data)
* Unbound logo. I added it as a webp file and a png file. But I can't remember how its logos are accessed for plugins. Also will need to add bg to the png.

**Comment by evanwporter** [comment]
image

Looks clean...

**Comment by evanwporter** [comment]
[80% unbound.zip](https://github.com/user-attachments/files/18194780/80.unbound.zip)

By the way been doing this save for tests which I found here => https://projectpokemon.org/home/files/file/4911-pokemon-unbound-save-file/

**Comment by evanwporter** [comment]
@andrewbenington 

Sprites are done

Now we just need to transfer restrictions & Items

**Review by evanwporter — COMMENTED** [review]
@andrewbenington

I finished the Unbound stuff. There may be a couple issues here and there with regards to the the correct sprites and the species name, but I they won't break anything and a tweak or two to the `conversions/` will resolve it in a jiffy.

Let me know what you think. 

When this is merged you should post OpenHome to `r/unbound` and `r/radicalred`. Mention how OpenHome can be used as an alternative to Unbound Cloud, except it has support for `Radical Red` & `Unbound` & every other mainline pokemon game (Unbound Cloud only supports Unbound).

**Comment by andrewbenington** [comment]
@evanwporter Looks good, just a couple image-related things. I don't want to add any new sprites to the codebase at this point, because it *really* takes up a lot of space and clogs up the git history. Even just moving the radical red sprites like you've done here has doubled the space those sprites take up in the .git files, because git is treating them like new files with their own history.

When I audited the git history to remove the old sprites it was a huge pain, but it really brought down the time/space needed to clone the repository. I don't want to have to do that again.

Right now my priority is getting a full release of the tauri version out, which should be any day now, but after that I will work on making it possible to download asset packs like Unbound sprites separately without them taking up space here. I would save the unbound sprites somewhere else on your computer to make it quick and easy to make them an asset pack when I get around to that.

I would love if more Radical Red/Unbound fans used OpenHome! But it should probably be you to post about it on those subreddits because I've never played them and I don't want to come off as a self-promoting poser 😂

File: public/logos/unbound.webp
Comment (andrewbenington) [inline public/logos/unbound.webp:1]: Is there a reason for having both a .webp and a .png version of this logo

File: src/pokemon/FileTypeSelect.tsx [hunk 1]
```diff
@@ -55,6 +55,8 @@ const FileTypeSelect = (props: FileTypeSelectProps) => {
   const [, , getEnabledSaveTypes] = useContext(AppInfoContext)
 
   const supportedFormats = useMemo(() => {
+    console.log(getEnabledSaveTypes())
```
Comment (andrewbenington) [inline src/pokemon/FileTypeSelect.tsx]: remove console log

**Review by andrewbenington — CHANGES_REQUESTED** [review]

**Comment by evanwporter** [comment]
@andrewbenington 

>Looks good, just a couple image-related things. I don't want to add any new sprites to the codebase at this point, because it really takes up a lot of space and clogs up the git history. Even just moving the radical red sprites like you've done here has doubled the space those sprites take up in the .git files, because git is treating them like new files with their own history.

A few images (less than 10) have been added but they take up a tiny bit. The 1000+ images you see in changed files are from when I changed the name `spites/rr` => `sprites/cfru`. Changed this back and the files shrunk from 1000+ files changed to 34 (at the time of writing). Also I've been making crazy changes left and right to the sprites. Which absolutely clog up the git history but this won't be an issue, as long as you click `squash and merge` and not just merge when this branch is merged. Unbound and rr pull from the same sprite folder.

**Comment by andrewbenington** [comment]
> A few images (less than 10) have been added but they take up a tiny bit. The 1000+ images you see in changed files are from when I changed the name `spites/rr` => `sprites/cfru`. Changed this back and the files shrunk from 1000+ files changed to 34 (at the time of writing). Also I've been making crazy changes left and right to the sprites. Which absolutely clog up the git history but this won't be an issue, as long as you click `squash and merge` and not just merge when this branch is merged. Unbound and rr pull from the same sprite folder.

@evanwporter Ah ok that's fine then. Btw in the future if you want to move or rename a file/directory, you can use `git mv <old name/location> <new name/location>` and git will consider it a rename instead of a deleted file and new file. That was helpful when I moved everything out of the old `renderer` folder. But yeah as long as it's squashed before the merge we don't have to worry about those commits on this branch

File: public/sprites/rr/.png
Comment (andrewbenington) [inline public/sprites/rr/.png:1]: This file should probably be deleted

**Comment by evanwporter** [comment]
@andrewbenington 

Last thing to do before we merge is the met location. The met location is registering as RadicalRed. Imo this is unavoidable since both roms have the same met-game metadata, so its either one or the other. Maybe someone smarter than me in the future will find a solution? If your okay with both roms occupying the same met game and met data then I think its ready to merge. Other issues can be resolved on a rolling basis like we did for #91.

File: src/types/SAVTypes/unbound/G3UBSAV.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+export class G3UBSAV extends G3CFRUSAV<PK3UB> implements PluginSAV<PK3UB> {
+  static transferRestrictions: TransferRestrictions = UB_TRANSFER_RESTRICTIONS
+
+  pluginIdentifier = 'unbound'
+
+  static saveTypeAbbreviation = 'Unbound'
+  static saveTypeName = 'Pokémon Unbound'
+  static saveTypeID = 'G3RRSAV'
```
Comment (andrewbenington) [inline src/types/SAVTypes/unbound/G3UBSAV.ts]:
```suggestion
  static saveTypeID = 'G3UBSAV'
```

File: src/types/SAVTypes/unbound/G3UBSAV.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+const UB_TRANSFER_RESTRICTIONS: TransferRestrictions = {
+  maxDexNum: NationalDex.Enamorus,
+  // excludedForms: {
+  // ...CapPikachus,
+  // [NationalDex.Pichu]: [SPIKY_EAR],
+  // [NationalDex.Dialga]: [ORIGIN],
+  // [NationalDex.Palkia]: [ORIGIN],
+  // },
+}
```
Comment (andrewbenington) [inline src/types/SAVTypes/unbound/G3UBSAV.ts]:
```suggestion
const UB_TRANSFER_RESTRICTIONS: TransferRestrictions = {
  maxDexNum: NationalDex.Enamorus,
}
```

File: src/types/SAVTypes/unbound/G3UBSAV.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    // const securityKey = bytesToUint32LittleEndian(firstSectionBytes, FRLG_SECURITY_OFFSET)
+    // const securityKeyCopy = bytesToUint32LittleEndian(firstSectionBytes, FRLG_SECURITY_COPY_OFFSET)
+
+    // console.log('UNBOUND SEC KEY', securityKey, securityKey === 0, securityKey !== securityKeyCopy)
+
+    // // Radical Red seems to have the security key set to 0, which has a 1 in 4.2 billion
+    // // chance to happen in vanilla FireRed (if it can even be 0 at all)
+    // return securityKey === 0 || securityKey !== securityKeyCopy
+  }
```
Comment (andrewbenington) [inline src/types/SAVTypes/unbound/G3UBSAV.ts]:
```suggestion
```

File: src/types/SAVTypes/cfru/PK3CFRU.ts [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+        this.metLocationIndex = other.metLocationIndex ?? FIRERED_IN_GAME_TRADE
+        this.internalMetLocationIndex = this.metLocationIndex
+      } else {
+        this.gameOfOrigin = other.gameOfOrigin ?? INTERNAL_ORIGIN_NON_RR
+        this.internalGameOfOrigin = INTERNAL_ORIGIN_NON_RR
+
+        this.metLocationIndex = other.metLocationIndex ?? FIRERED_IN_GAME_TRADE
+        this.internalMetLocationIndex = FIRERED_IN_GAME_TRADE
+      }
+
```
Comment (andrewbenington) [inline src/types/SAVTypes/cfru/PK3CFRU.ts:297]: I would move this code out of the super class and into the subclasses, after a `super()` call. This is why you're having issues tracking the `unbound` origin

File: src/types/SAVTypes/cfru/PK3CFRU.ts [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+      // 51:53
+      this.metLevel = uIntFromBufferBits(dataView, 0x34, 0, 7, true)
+
+      this.gameOfOrigin = uIntFromBufferBits(dataView, 0x34, 7, 4, true)
+      this.internalGameOfOrigin = this.gameOfOrigin
+
+      // OHPKM will not copy this field if the mon was already being tracked before being
+      // transferred to Radical Red
+      this.pluginOrigin = this.gameOfOrigin === INTERNAL_ORIGIN_FROM_RR ? 'radical_red' : undefined
```
Comment (andrewbenington) [inline src/types/SAVTypes/cfru/PK3CFRU.ts]: same here

**Review by andrewbenington — DISMISSED** [review]
There are also a bunch of other commented out sections of code I assume you can remove.

If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.

**Comment by evanwporter** [comment]
>If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.

Well what if a mon originates in RR but is moved to Unbound, wouldn't it then say its from RR

**Comment by andrewbenington** [comment]
> > If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.
> 
> Well what if a mon originates in RR but is moved to Unbound, wouldn't it then say its from Unbound

Nope, that mon's OHPKM file will already have its pluginOrigin set to `radical_red` from when it was moved out of Radical Red. The `pluginOrigin` field is only set once, and never updated

**Comment by evanwporter** [comment]
Oh okay. Yeah how might that work then?

File: src/types/SAVTypes/unbound/G3UBSAV.ts (hunk 2 above)
Comment (evanwporter) [reply src/types/SAVTypes/unbound/G3UBSAV.ts]: Really good catch here!

**Comment by evanwporter** [comment]
@andrewbenington 

Nvm I got it to work. Looks clean.

image

**Comment by andrewbenington** [comment]
> @andrewbenington
> 
> Nvm I got it to work. Looks clean.
> 
> image

Looks great. I like the dropdown color you picked too.

I'm going to do some testing on my own in the coming days and then I'll merge this. Apologies in advance if this takes a bit; I'm pretty busy this week with the holidays

**Review by andrewbenington — APPROVED** [review]

**Pull request merged**